File: cli/core/decks.go
// Listagem de decks (spec/SPEC.md §4.5) — RF07.
// Deck = unidade de estudo com nome e tags transversais; hierarquia é
// convenção de nome com "::" (derivada na exibição, não aqui).
package core

type Deck struct {
	Name     string   `json:"name"`
	Tags     []string `json:"tags"`
	Archived bool     `json:"archived"`
}

type DecksState struct {
	Decks map[string]Deck `json:"decks"`
}

func init() {
	Register("decks", func(events []Event, _ Params) (any, error) {
		return Decks(events)
	})
}

func Decks(events []Event) (DecksState, error) {
	state := DecksState{Decks: map[string]Deck{}}

	for _, event := range Normalize(events) {
		switch event.Type {
		case "deck.created":
			pl, err := decodePayload[struct {
				DeckID string   `json:"deck_id"`
				Name   string   `json:"name"`
				Tags   []string `json:"tags"`
			}](event)
			if err != nil {
				return DecksState{}, err
			}
			state.Decks[pl.DeckID] = Deck{Name: pl.Name, Tags: orEmpty(pl.Tags)}

		case "deck.renamed":
			pl, err := decodePayload[struct {
				DeckID string `json:"deck_id"`
				Name   string `json:"name"`
			}](event)
			if err != nil {
				return DecksState{}, err
			}
			if deck, ok := state.Decks[pl.DeckID]; ok {
				deck.Name = pl.Name // último nome vence
				state.Decks[pl.DeckID] = deck
			}

		case "deck.archived":
			pl, err := decodePayload[struct {
				DeckID string `json:"deck_id"`
			}](event)
			if err != nil {
				return DecksState{}, err
			}
			if deck, ok := state.Decks[pl.DeckID]; ok {
				deck.Archived = true
				state.Decks[pl.DeckID] = deck
			}
		}
	}
	return state, nil
}

File: extension/core/decks.js
// Listagem de decks (spec/SPEC.md §3, Estudo) — RF07.
// Deck = unidade de estudo com nome e tags transversais.

import { normalize } from "./envelope.js";

export function project(events) {
  const decks = {};

  for (const event of normalize(events)) {
    if (event.type === "deck.created") {
      const { deck_id, name, tags } = event.payload;
      decks[deck_id] = { name, tags: tags ?? [], archived: false };
    } else if (event.type === "deck.renamed") {
      const deck = decks[event.payload.deck_id];
      if (deck) deck.name = event.payload.name; // último nome vence
    } else if (event.type === "deck.archived") {
      const deck = decks[event.payload.deck_id];
      if (deck) deck.archived = true;
    }
  }

  return { decks };
}
